### PR TITLE
Dev create definition

### PR DESCRIPTION
**Pull request by RodPaDev**
Alright, so there was a lot figuring out and there is so much more. I've got to say you got one impressing system and I may understand less than 1% of it.

So I managed to create the definition, create the directory, write the JSON and copy the OBJ from the pack. 
It "shows" up in the scenery tab after reloading the page, the thing is the model doesn't load and I get an Invalid model log.

Now the code is by no means up to your standard that I can say. I will still open a PR for it and will still keep on working to make it work. Needless to say, it took me way over 30 minutes.

Anyways, I'm not satisfied with this and totally accept if you deny my application

### CODE REVIEW

File: server/tools/scenery.js [hunk 1]
```diff
@@ -87,9 +87,34 @@ function deleteModel(workspace, body) {
     return true;
 }
 
-function createDefinition(workspace, body) {
-    console.log("Create Definition: " + workspace + ": " + json(body));
-    return false;
+function createDefinition(workspace, body) { 
+    let objects = JSON.parse(fs.readFileSync(root_dir + workspace + '/objects.json'));
```
Comment (Heleor) [inline server/tools/scenery.js]: You shouldn't need 'objects' in order to do this.

File: server/tools/scenery.js [hunk 2]
```diff
@@ -87,9 +87,34 @@ function deleteModel(workspace, body) {
...
+
+
+    const workspacePath = root_dir + workspace
```
Comment (Heleor) [inline server/tools/scenery.js]: Nit: I generally prefer to have semicolons ending every line, even if they're not required.

File: server/tools/scenery.js [hunk 3]
```diff
@@ -87,9 +87,34 @@ function deleteModel(workspace, body) {
...
+    const modelPath = "./assets/" + body.pack + "/" + body.model
+    const definitionPath = "/models/definitions/" + body.id.split("-").slice(0,-1).join("/") + "/";
```
Comment (Heleor) [inline server/tools/scenery.js]: I like doing:

let pieces = body.id.split('-');
let name = pieces.pop(); // or shift, I get them mixed up
let path = pieces.join('/'); 

This way you get both the filename and folder path.

File: server/tools/scenery.js [hunk 4]
```diff
@@ -87,9 +87,34 @@ function deleteModel(workspace, body) {
...
+    const path = workspacePath + definitionPath
+
+
+    delete body.pack // no longer needed
```
Comment (Heleor) [inline server/tools/scenery.js]: No reason to do this.

File: server/tools/scenery.js [hunk 5]
```diff
@@ -87,9 +87,34 @@ function deleteModel(workspace, body) {
...
+
+    if(!fs.existsSync(path)){
```
Comment (Heleor) [inline server/tools/scenery.js]: Use fs.ensureDirSync() instead.

File: server/tools/scenery.js [hunk 6]
```diff
@@ -87,9 +87,34 @@ function deleteModel(workspace, body) {
...
+        fs.mkdirSync(path, {recursive: true})
+    }
+
+    const fileName = body.model.match(/(?=\w+\.\w{3,4}$).+/)[0];
```
Comment (Heleor) [inline server/tools/scenery.js]: If you look at the comments on line 96 you don't have to do this.

File: server/tools/scenery.js [hunk 7]
```diff
@@ -87,9 +87,34 @@ function deleteModel(workspace, body) {
...
+
+    body.model = fileName;
+
+    const jsonFileName = body.model.match(/(?=\w+\.\w{3,4}$).+/)[0].replace(".obj", ".json");
```
Comment (Heleor) [inline server/tools/scenery.js]: Ew. Avoid regex.

**Review by Heleor — COMMENTED** [review]
This is actually mostly right. The reason why the models you make don't show up are because you didn't copy the texture over into shared-textures.

File: server/tools/scenery.js (hunk 3 above)
Comment (RodPaDev) [reply server/tools/scenery.js]: Oh nice didn't think of it

File: server/tools/scenery.js (hunk 2 above)
Comment (RodPaDev) [reply server/tools/scenery.js]: I feel embarrassed, I mostly code in c# lately, how could I

File: server/tools/scenery.js (hunk 7 above)
Comment (RodPaDev) [reply server/tools/scenery.js]: It was a questionable choice I agree

File: server/tools/scenery.js [hunk 8]
```diff
@@ -87,9 +87,40 @@ function deleteModel(workspace, body) {
...
+    // this will also rename the original .obj file when copying
+    body.model = name + ".obj";
+
+    fs.ensureDirSync(workspacePath + definitionPath + path);
+
+    const changes = body.changes;
+
+    delete body.changes
+    delete body.id
+    delete body.pack;
```
Comment (Heleor) [inline server/tools/scenery.js]: 99% of the time, if you're typing a 'delete' operation you're making a mistake. You should never have to delete any keys.

File: server/tools/scenery.js [hunk 9]
```diff
@@ -87,9 +87,40 @@ function deleteModel(workspace, body) {
...
+
+    fs.writeFileSync(workspacePath + definitionPath + path + name + ".json", json(merge(body, changes)));
```
Comment (Heleor) [inline server/tools/scenery.js]: Merge should take a base object that you define, not body.

```
let def = {
  sharedTexture: body.sharedTexture,
  model: body.model,
}
let definition = merge(def, changes);
fs.writeFileSync(..., json(definition));
```

File: client/models/model_editor.js [hunk 10]
```diff
@@ -285,15 +285,17 @@ class Models {
...
             let texture = this.asset_definition.texture;
+            let pack = this.loaded_asset_pack
 
             post('api/tools/scenery/definition/create/' + WORKSPACES.opened, {
                 id: id,
                 model: model,
                 sharedTexture: texture,
-                changes: changes
+                changes: changes,
+                pack: pack, // I added the pack because I'm not sure if this is what you wanted, it's deleted before writing json
```
Comment (Heleor) [inline client/models/model_editor.js]: Delete the comment.

**Pull request merged**